Bitcoind backend does not work following the docs

I'm attempting to setup bitcoind backend following the docs here:

https://bitcoin-s.org/docs/next/getting-setup#option-b-bitcoind-backend

```
    bitcoind-rpc {
        # bitcoind rpc username
        rpcuser = my_user_name
        # bitcoind rpc password
        rpcpassword = my_password

        # Binary location of bitcoind
        binary = ${HOME}/.bitcoin-s/binaries/bitcoind/bitcoin-0.21.0/bin/bitcoind
        # bitcoind datadir
        datadir = ${HOME}/.bitcoin
        # bitcoind network binding
        bind = localhost
        # bitcoind p2p port
        port = 8333
        # bitcoind rpc binding
        rpcbind = localhost
        # bitcoind rpc port
        rpcport = 8332
        # bitcoind zmq port for all services
        zmqport = 29000
}
```

my `bitcoin.conf`

```
addresstype=bech32
server=1
txindex=1
daemon=1
dbcache=4096
blockfilterindex=1
peerblockfilters=1

rpcuser=...
rpcpassword=...
zmqpubhashtx=tcp://127.0.0.1:29000
zmqpubashblock=tcp://127.0.0.1:29000
zmqpubrawblock=tcp://127.0.0.1:29000
zmqpubrawtx=tcp://127.0.0.1:29000
```


and the error I get when running

>2021-04-12T12:28:18UTC INFO [BitcoinSServerMain] Starting wallet
2021-04-12T12:28:18UTC INFO [Wallet$WalletImpl] Starting Wallet
2021-04-12T12:28:18UTC INFO [DbValidate] Successfully validated 10 migrations (execution time 00:00.005s)
2021-04-12T12:28:18UTC INFO [DbMigrate] Current version of schema "main": 10
2021-04-12T12:28:18UTC INFO [DbMigrate] Schema "main" is up to date. No migration necessary.
2021-04-12T12:28:44UTC INFO [ZMQSubscriber] starting zmq
2021-04-12T12:28:44UTC INFO [ZMQSubscriber] ZmqSubscriber connecting to uri=tcp://127.0.0.1:29000
Exception in thread "ZMQSubscriber-thread-1618230524433" org.zeromq.ZMQException: tcp://127.0.0.1 : errno 49
        at zmq.io.net.tcp.TcpAddress.resolve(TcpAddress.java:155)
        at zmq.io.net.tcp.TcpAddress.<init>(TcpAddress.java:39)
        at zmq.io.net.Address.resolve(Address.java:93)
        at zmq.SocketBase.connect(SocketBase.java:549)
        at org.zeromq.ZMQ$Socket.connect(ZMQ.java:3082)
        at org.bitcoins.zmq.ZMQSubscriber$SubscriberRunnable$.run(ZMQSubscriber.scala:43)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.net.UnknownHostException: tcp://127.0.0.1
        at java.base/java.net.InetAddress$CachedAddresses.get(InetAddress.java:797)
        at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1505)
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1364)
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1298)
        at zmq.io.net.tcp.TcpAddress.resolve(TcpAddress.java:132)
        ... 6 more

--- a/app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala
@@ -18,6 +18,7 @@ import org.bitcoins.feeprovider._
 import org.bitcoins.node._
 import org.bitcoins.node.config.NodeAppConfig
 import org.bitcoins.node.models.Peer
+import org.bitcoins.rpc.config.ZmqConfig
 import org.bitcoins.server.routes.{BitcoinSRunner, Server}
 import org.bitcoins.wallet.Wallet
 import org.bitcoins.wallet.config.WalletAppConfig
@@ -147,15 +148,14 @@ class BitcoinSServerMain(override val args: Array[String])
 
         blockCount <- bitcoind.getBlockCount
         // Create callbacks for processing new blocks
-        _ = bitcoindRpcConf.zmqPortOpt match {
-          case Some(_) =>
-            val zmq = BitcoindRpcBackendUtil.createZMQWalletCallbacks(wallet)
-            zmq.start()
-          case None =>
+        _ =
+          if (bitcoindRpcConf.zmqConfig == ZmqConfig()) {
             BitcoindRpcBackendUtil.startBitcoindBlockPolling(wallet,
                                                              bitcoind,
                                                              blockCount)
-        }
+          } else {
+            BitcoindRpcBackendUtil.startZMQWalletCallbacks(wallet)
+          }
 
         binding <- startHttpServer(nodeApi = bitcoind,
                                    chainApi = bitcoind,

--- a/app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/BitcoinSServerMain.scala
@@ -149,7 +149,7 @@ class BitcoinSServerMain(override val args: Array[String])
         blockCount <- bitcoind.getBlockCount
         // Create callbacks for processing new blocks
         _ =
-          if (bitcoindRpcConf.zmqConfig == ZmqConfig()) {
+          if (bitcoindRpcConf.zmqConfig == ZmqConfig.empty) {
             BitcoindRpcBackendUtil.startBitcoindBlockPolling(wallet,
                                                              bitcoind,
                                                              blockCount)

--- a/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcAppConfig.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcAppConfig.scala
@@ -1,9 +1,5 @@
 package org.bitcoins.server
 
-import java.io.File
-import java.net.URI
-import java.nio.file._
-
 import akka.actor.ActorSystem
 import com.typesafe.config.Config
 import org.bitcoins.db._
@@ -12,6 +8,9 @@ import org.bitcoins.node.config.NodeAppConfig
 import org.bitcoins.rpc.client.common.BitcoindRpcClient
 import org.bitcoins.rpc.config._
 
+import java.io.File
+import java.net.{InetSocketAddress, URI}
+import java.nio.file._
 import scala.concurrent.{ExecutionContext, Future}
 
 /** Configuration for a BitcoindRpcClient
@@ -91,13 +90,32 @@ case class BitcoindRpcAppConfig(
   lazy val authCredentials: BitcoindAuthCredentials =
     BitcoindAuthCredentials.PasswordBased(rpcUser, rpcPassword)
 
-  lazy val zmqPortOpt: Option[Int] =
-    config.getIntOpt("bitcoin-s.bitcoind-rpc.zmqport")
+  lazy val zmqRawBlock: Option[InetSocketAddress] =
+    config.getStringOrNone("bitcoin-s.bitcoind-rpc.zmqpubrawblock").map { str =>
+      val uri = URI.create(str)
+      new InetSocketAddress(uri.getHost, uri.getPort)
+    }
 
-  lazy val zmqConfig: ZmqConfig = zmqPortOpt match {
-    case Some(zmqPort) => ZmqConfig.fromPort(zmqPort)
-    case None          => ZmqConfig()
-  }
+  lazy val zmqRawTx: Option[InetSocketAddress] =
+    config.getStringOrNone("bitcoin-s.bitcoind-rpc.zmqpubrawtx").map { str =>
+      val uri = URI.create(str)
+      new InetSocketAddress(uri.getHost, uri.getPort)
+    }
+
+  lazy val zmqHashBlock: Option[InetSocketAddress] =
+    config.getStringOrNone("bitcoin-s.bitcoind-rpc.zmqpubashblock").map { str =>
+      val uri = URI.create(str)
+      new InetSocketAddress(uri.getHost, uri.getPort)
+    }
+
+  lazy val zmqHashTx: Option[InetSocketAddress] =
+    config.getStringOrNone("bitcoin-s.bitcoind-rpc.zmqpubashtx").map { str =>
+      val uri = URI.create(str)
+      new InetSocketAddress(uri.getHost, uri.getPort)
+    }
+
+  lazy val zmqConfig: ZmqConfig =
+    ZmqConfig(zmqHashBlock, zmqRawBlock, zmqHashTx, zmqRawTx)
 
   lazy val bitcoindInstance: BitcoindInstance =
     BitcoindInstance(network = network,

--- a/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcBackendUtil.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcBackendUtil.scala
@@ -106,7 +106,7 @@ object BitcoindRpcBackendUtil extends Logging {
 
   def startZMQWalletCallbacks(wallet: Wallet)(implicit
       bitcoindRpcConf: BitcoindRpcAppConfig): Unit = {
-    require(bitcoindRpcConf.zmqConfig != ZmqConfig(),
+    require(bitcoindRpcConf.zmqConfig != ZmqConfig.empty,
             "Must have the zmq raw configs defined to setup ZMQ callbacks")
 
     bitcoindRpcConf.zmqRawTx.foreach { zmq =>

--- a/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcBackendUtil.scala
+++ b/app/server/src/main/scala/org/bitcoins/server/BitcoindRpcBackendUtil.scala
@@ -1,18 +1,17 @@
 package org.bitcoins.server
 
 import akka.actor.{ActorSystem, Cancellable}
+import grizzled.slf4j.Logging
 import org.bitcoins.core.api.node.NodeApi
 import org.bitcoins.core.protocol.blockchain.Block
 import org.bitcoins.core.protocol.transaction.Transaction
 import org.bitcoins.core.util.FutureUtil
-import grizzled.slf4j.Logging
-
 import org.bitcoins.crypto.DoubleSha256Digest
 import org.bitcoins.rpc.client.common.BitcoindRpcClient
+import org.bitcoins.rpc.config.ZmqConfig
 import org.bitcoins.wallet.Wallet
 import org.bitcoins.zmq.ZMQSubscriber
 
-import java.net.InetSocketAddress
 import java.util.concurrent.atomic.AtomicReference
 import scala.concurrent.duration.{DurationInt, FiniteDuration}
 import scala.concurrent.{ExecutionContext, Future, Promise}
@@ -105,34 +104,43 @@ object BitcoindRpcBackendUtil extends Logging {
     pairedWallet
   }
 
-  def createZMQWalletCallbacks(wallet: Wallet)(implicit
-      bitcoindRpcConf: BitcoindRpcAppConfig): ZMQSubscriber = {
-    require(bitcoindRpcConf.zmqPortOpt.isDefined,
-            "Must have the zmq port defined to setup ZMQ callbacks")
-    val zmqSocket =
-      new InetSocketAddress("tcp://127.0.0.1", bitcoindRpcConf.zmqPortOpt.get)
+  def startZMQWalletCallbacks(wallet: Wallet)(implicit
+      bitcoindRpcConf: BitcoindRpcAppConfig): Unit = {
+    require(bitcoindRpcConf.zmqConfig != ZmqConfig(),
+            "Must have the zmq raw configs defined to setup ZMQ callbacks")
 
-    val rawTxListener: Option[Transaction => Unit] = Some {
-      { tx: Transaction =>
-        logger.debug(s"Received tx ${tx.txIdBE}, processing")
-        wallet.processTransaction(tx, None)
-        ()
+    bitcoindRpcConf.zmqRawTx.foreach { zmq =>
+      val rawTxListener: Option[Transaction => Unit] = Some {
+        { tx: Transaction =>
+          logger.debug(s"Received tx ${tx.txIdBE.hex}, processing")
+          wallet.processTransaction(tx, None)
+          ()
+        }
       }
+
+      new ZMQSubscriber(socket = zmq,
+                        hashTxListener = None,
+                        hashBlockListener = None,
+                        rawTxListener = rawTxListener,
+                        rawBlockListener = None).start()
     }
 
-    val rawBlockListener: Option[Block => Unit] = Some {
-      { block: Block =>
-        logger.debug(s"Received block ${block.blockHeader.hashBE}, processing")
-        wallet.processBlock(block)
-        ()
+    bitcoindRpcConf.zmqRawBlock.foreach { zmq =>
+      val rawBlockListener: Option[Block => Unit] = Some {
+        { block: Block =>
+          logger.debug(
+            s"Received block ${block.blockHeader.hashBE.hex}, processing")
+          wallet.processBlock(block)
+          ()
+        }
       }
-    }
 
-    new ZMQSubscriber(socket = zmqSocket,
-                      hashTxListener = None,
-                      hashBlockListener = None,
-                      rawTxListener = rawTxListener,
-                      rawBlockListener = rawBlockListener)
+      new ZMQSubscriber(socket = zmq,
+                        hashTxListener = None,
+                        hashBlockListener = None,
+                        rawTxListener = None,
+                        rawBlockListener = rawBlockListener).start()
+    }
   }
 
   private def getNodeApiWalletCallback(

--- a/bitcoind-rpc/src/main/scala/org/bitcoins/rpc/config/ZmqConfig.scala
+++ b/bitcoind-rpc/src/main/scala/org/bitcoins/rpc/config/ZmqConfig.scala
@@ -13,6 +13,8 @@ sealed trait ZmqConfig {
 
 object ZmqConfig extends Logging {
 
+  val empty: ZmqConfig = ZmqConfig()
+
   private case class ZmqConfigImpl(
       hashBlock: Option[InetSocketAddress],
       rawBlock: Option[InetSocketAddress],

--- a/docs/config/configuration.md
+++ b/docs/config/configuration.md
@@ -144,6 +144,10 @@ bitcoin-s {
         rpcport = 8332
         # bitcoind zmq port for all services
         zmqport = 29000
+        # bitcoind zmq raw tx
+        zmqpubrawtx = "tcp://127.0.0.1:28332"
+        # bitcoind zmq raw block
+        zmqpubrawblock = "tcp://127.0.0.1:28333"
     }
 
     node {

--- a/docs/config/configuration.md
+++ b/docs/config/configuration.md
@@ -148,6 +148,10 @@ bitcoin-s {
         zmqpubrawtx = "tcp://127.0.0.1:28332"
         # bitcoind zmq raw block
         zmqpubrawblock = "tcp://127.0.0.1:28333"
+        # bitcoind zmq hash tx
+        zmqpubhashtx = "tcp://127.0.0.1:28330"
+        # bitcoind zmq raw block
+        zmqpubhashblock = "tcp://127.0.0.1:28331"
     }
 
     node {

--- a/docs/getting-setup.md
+++ b/docs/getting-setup.md
@@ -153,8 +153,10 @@ bitcoin-s {
         rpcbind = localhost
         # bitcoind rpc port
         rpcport = 8332
-        # bitcoind zmq port for all services
-        zmqport = 29000
+        # bitcoind zmq raw tx
+        zmqpubrawtx = "tcp://127.0.0.1:28332"
+        # bitcoind zmq raw block
+        zmqpubrawblock = "tcp://127.0.0.1:28333"
     }
 ```
 

